pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -271,29 +271,30 @@ func BuildHTTPRoutesForVirtualService(
 
 	out := make([]*route.Route, 0, len(vs.Http))
 
-	catchall := false
 	for _, http := range vs.Http {
 		if len(http.Match) == 0 {
-			if r := translateRoute(push, node, http, nil, listenPort, virtualService, serviceRegistry, gatewayNames); r != nil {
+			r := translateRoute(push, node, http, nil, listenPort, virtualService, serviceRegistry, gatewayNames)
+			if r != nil {
 				out = append(out, r)
 			}
-			catchall = true
-		} else {
-			for _, match := range http.Match {
-				if r := translateRoute(push, node, http, match, listenPort, virtualService, serviceRegistry, gatewayNames); r != nil {
+			return out, nil
+		}
+
+		tmp := []*route.Route{}
+		for _, match := range http.Match {
+			if r := translateRoute(push, node, http, match, listenPort, virtualService, serviceRegistry, gatewayNames); r != nil {
+				tmp = append(tmp, r)
+				// This is a catch all path. Routes are matched in order, so we will never go beyond this match
```

Comment:
I vaguely recall we used to do this and reverted it for some reason.. let me see if I can dig up the history

Reply:
https://github.com/istio/istio/pull/27449 was what I was thinking of

Reply:
HTTPRoute.Match
	// Match conditions to be satisfied for the rule to be
	// activated. All conditions inside a single match block have AND
	// semantics, while the list of match blocks have OR semantics. The rule
	// is matched if any one of the match blocks succeed.

This is the api doc, it says it matches when anyone condition matches. If there is a match all, all other routes are redundant

Reply:
@howardjohn After a close look at the issue, it is different, with this pr it does not remove the `provider-global-default` catch all route. There is ut for this. And i am testing manually in my env

Reply:
Verified it does not remove the catchall `provider-global-default`
```
[
    {
        "name": "8000",
        "virtualHosts": [
            {
                "name": "allow_any",
                "domains": [
                    "*"
                ],
                "routes": [
                    {
                        "name": "allow_any",
                        "match": {
                            "prefix": "/"
                        },
                        "route": {
                            "cluster": "PassthroughCluster",
                            "timeout": "0s",
                            "maxGrpcTimeout": "0s"
                        }
                    }
                ],
                "includeRequestAttemptCount": true
            },
            {
                "name": "httpbin.default.svc.cluster.local:8000",
                "domains": [
                    "httpbin.default.svc.cluster.local",
                    "httpbin.default.svc.cluster.local:8000",
                    "httpbin",
                    "httpbin:8000",
                    "httpbin.default.svc",
                    "httpbin.default.svc:8000",
                    "httpbin.default",
                    "httpbin.default:8000",
                    "10.96.196.41",
                    "10.96.196.41:8000"
                ],
                "routes": [
                    {
                        "name": "GET-/",
                        "match": {
                            "prefix": "/",
                            "caseSensitive": true,
                            "headers": [
                                {
                                    "name": ":method",
                                    "exactMatch": "GET"
                                }
                            ]
                        },
                        "route": {
                            "cluster": "outbound|8000||httpbin.default.svc.cluster.local",
                            "timeout": "4s",
                            "retryPolicy": {
                                "retryOn": "gateway-error,connect-failure,retriable-status-codes",
                                "numRetries": 3,
                                "perTryTimeout": "1s",
                                "retryHostPredicate": [
                                    {
                                        "name": "envoy.retry_host_predicates.previous_hosts"
                                    }
                                ],
                                "hostSelectionRetryMaxAttempts": "5"
                            },
                            "maxGrpcTimeout": "4s"
                        },
                        "metadata": {
                            "filterMetadata": {
                                "istio": {
                                    "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bar"
                                }
                            }
                        },
                        "decorator": {
                            "operation": "httpbin.default.svc.cluster.local:8000/*"
                        },
                        "requestHeadersToAdd": [
                            {
                                "header": {
                                    "key": "x-envoy-retriable-status-codes",
                                    "value": "408,423"
                                },
                                "append": true
                            }
                        ]
                    },
                    {
                        "name": "provider-global-default",
                        "match": {
                            "prefix": "/"
                        },
                        "route": {
                            "cluster": "outbound|8000||httpbin.default.svc.cluster.local",
                            "timeout": "0s",
                            "retryPolicy": {
                                "retryOn": "connect-failure,refused-stream,unavailable,cancelled,retriable-status-codes",
                                "numRetries": 2,
                                "retryHostPredicate": [
                                    {
                                        "name": "envoy.retry_host_predicates.previous_hosts"
                                    }
                                ],
                                "hostSelectionRetryMaxAttempts": "5",
                                "retriableStatusCodes": [
                                    503
                                ]
                            },
                            "maxGrpcTimeout": "0s"
                        },
                        "metadata": {
                            "filterMetadata": {
                                "istio": {
                                    "config": "/apis/networking.istio.io/v1alpha3/namespaces/default/virtual-service/bar"
                                }
                            }
                        },
                        "decorator": {
                            "operation": "httpbin.default.svc.cluster.local:8000/*"
                        }
                    }
                ],
                "includeRequestAttemptCount": true
            }
        ],
        "validateClusters": false
    }
]

```

Reply:
Thanks for looking into it!